src/tokenizer.jl

```diff
@@ -61,7 +61,84 @@ function Tokenizer(tokenizer_path::String, vocab_size::Int)
     return Tokenizer(tokens, vocab_scores)
 end
 
-const BOS_TOKEN::Int32 = 1
+const BOS_TOKEN::Int32 = 2
+const EOS_TOKEN::Int32 = 3
+
+"""
+Encode the string text (input) into an upper-bound preallocated token array
+bos != 0 means prepend the BOS token (=1), eos != 0 means append the EOS token (=2)
+
+llama2.c correspondence: encode (l. 452)
+"""
+function encode(tokenizer::Tokenizer, text::String)
+    # stores merge candidates of two consecutive tokens
+    tokens::Vector{Int} = []
+
+    # if the text is empty there is no encoding so we can return the empty tokens array
+    if isempty(text)
+        return tokens
+    end
+
+    # add BOS token
+    push!(tokens, BOS_TOKEN)
+    # add dummy_prefix is default
+    # add prefix to the input string only if text isn't empty
+    # not sure why though
+    if text != ""
+        dummy_prefix::Int32 = tokenizer.token_to_index[" "]
+        push!(tokens, dummy_prefix)
+    end
+
+    # process the bytes of the text
+    # no need for str_buffers (as in llama2.c) because of julias native codeunits
+    for c in codeunits(text)
+        # add token of the byte to the tokens list
+        token = String(UInt8[c])
+        push!(tokens, tokenizer.token_to_index[token])
+    end
+
+    # merge the best consecutive pair each iteration
+    while true
+        best_score::Float32 = -1e10
+        best_id::Int32 = -1
+        best_idx::Int32 = -1
+
+        for i in 1:(length(tokens) - 1)
+            # merge consecutive tokens 
+            merged_token =
+                tokenizer.index_to_token[tokens[i]] *
+                tokenizer.index_to_token[tokens[i+1]]
+
+            # check if merged_token exists
+            id = get(tokenizer.token_to_index, merged_token, nothing)
+
+            # if id exists and the vocab_score is bigger than the best score, this token becomes the new best token
+            if !isnothing(id) && tokenizer.vocab_scores[id] > best_score
+                best_score = tokenizer.vocab_scores[id]
+                best_id = id
+                best_idx = i
+            end
+        end
+
+        # if no more tokens can be merged, we break
+        if best_idx == -1
+            break
+        end
+
+        # merge consecutive pair into new token best_id
+        tokens[best_idx] = best_id
+
+        # delete token at position best_idx+1, shift entire sequence back 1
+        splice!(tokens, best_idx + 1)
+    end
+
+    # add EOS token
+    # is it place 2? Because Julia starts counting at 1
+    push!(tokens, EOS_TOKEN)
```

Comment:
Looking at run.c, the EOS token is always set to zero there. So EOS is never actually added. Either remove this functionality or make it optional (by providing a boolean parameter with a default value of false).
+ Remove `# is it place 2? ..`